stellargraph/layer/watch_your_step.py

```diff
@@ -0,0 +1,202 @@
+
+        return expected_walk
+
+
+@experimental(reason="lack of unit tests")
+class WatchYourStep:
+    """
+    Implementation of the node embeddings as in Watch Your Step: Learning Node Embeddings via Graph Attention
+    https://arxiv.org/pdf/1710.09599.pdf.
+
+    This model requires specification of the number of random walks starting from each node, and the embedding dimension
+    to use for the node embeddings. Note, that the embedding dimension should be an even number or else will be
+    rounded down to nearest even number.
+
+    Args:
+        generator (AdjacencyPowerGenerator): the generator
+        num_walks (int): the number of random walks starting at each node to use when calculating the expected random
+            walks. Defaults to `80` as this value was found to perform well by the authors of the paper.
+        embedding dimension (int): the dimension to use for the node embeddings. Defaults to `64`
+            as this value was found to perform well by the authors of the paper.
+    """
+
+    def __init__(
+        self,
+        generator,
+        num_walks=80,
+        embedding_dimension=64,
+        attention_regularizer=None,
+        attention_initializer=None,
+        attention_constraint=None,
+    ):
+
+        if not isinstance(generator, AdjacencyPowerGenerator):
+            raise TypeError(
+                "generator should be an instance of AdjacencyPowerGenerator."
+            )
+
+        if not isinstance(num_walks, int):
+            raise TypeError("num_walks should be an int.")
+
+        if num_walks <= 0:
+            raise ValueError("num_walks should be a positive int.")
+
+        self.num_walks = num_walks
+        self.num_powers = generator.num_powers
+        self.n_nodes = int(generator.Aadj_T.shape[0])
+        self.embedding_dimension = embedding_dimension
+        self.attention_regularizer = attention_regularizer
+        self.attention_initializer = attention_initializer
+        self.attention_constraint = attention_constraint
+
+    def build(self):
+        """
+        This function builds the layers for a keras model.
+
+        returns:
+            A tuple of (inputs, outputs) to use with a keras model.
+        """
+
+        input_rows = Input(batch_shape=(None,), name="row_node_ids", dtype="int64")
+        input_powers = Input(batch_shape=(None, self.num_powers, self.n_nodes))
+
+        left_embedding = Embedding(
+            self.n_nodes,
+            self.embedding_dimension,
+            input_length=None,
+            name="WATCH_YOUR_STEP_LEFT_EMBEDDINGS",
+        )
+
+        right_embedding = Embedding(
+            self.n_nodes,
+            self.embedding_dimension,
+            input_length=None,
+            name="WATCH_YOUR_STEP_RIGHT_EMBEDDINGS",
+        )
+
+        vectors_left = left_embedding(input_rows)
+
+        # TODO: replace the embedding layer with a custom layer to avoid lookups
```

Comment:
Does it have to be custom? We could potentially model it as a trainable `n × d/2` matrix that is multiplied with the left embeddings (i.e. a dense layer). Does that make sense? What do you think?

Reply:
That's exactly what I was thinking of doing but wrapped in a custom keras layer (all trainable weights have to be in keras layers). Although maybe a `Dense` layer will do the trick here...

Reply:
nice! the dense layer worked